src/MizarModels/include/MizarModels/SamplingWithFrameTrackReportModel.h

```diff
@@ -104,31 +113,63 @@ class SamplingWithFrameTrackReportModelTmpl : public QAbstractTableModel {
   }
 
  private:
+  struct Index {
+    SFID sfid;
+    Column column;
+  };
+
   void EmitDataChanged(Column column) {
     const int column_int = static_cast<int>(column);
     emit dataChanged(index(0, column_int), index(rowCount() - 1, column_int));
   }
 
-  [[nodiscard]] std::string MakeDisplayedString(const QModelIndex& index) const {
-    const SFID sfid = sfids_[index.row()];
-    const auto column = static_cast<Column>(index.column());
-    switch (column) {
-      case Column::kFunctionName:
-        return report_.GetSfidToNames().at(sfid);
+  [[nodiscard]] Index MakeIndex(const QModelIndex& index) const {
+    return {sfids_[index.row()], static_cast<Column>(index.column())};
+  }
+
+  [[nodiscard]] bool IsNumeric(const Index& index, int role) const {
+    switch (index.column) {
       case Column::kBaselineExclusivePercent:
       case Column::kBaselineExclusiveTimePerFrame:
       case Column::kComparisonExclusivePercent:
       case Column::kComparisonExclusiveTimePerFrame:
       case Column::kPvalue:
       case Column::kSlowdownPercent:
       case Column::kPercentOfSlowdown:
-        return absl::StrFormat("%.3f", MakeNumericEntry(sfid, column));
+        return true;
+      case Column::kIsSignificant:
+        return role != Qt::DisplayRole;
+      default:
+        return false;
+    }
+  }
+
+  [[nodiscard]] QVariant MakeSortValue(const QModelIndex& model_index) const {
+    const Index index = MakeIndex(model_index);
+
+    if (IsNumeric(index, kSortRole)) return MakeNumericEntry(index);
+    return QString::fromStdString(MakeStringEntry(index));
+  }
+
+  [[nodiscard]] std::string MakeDisplayedString(const QModelIndex& model_index) const {
+    const Index index = MakeIndex(model_index);
+
+    if (IsNumeric(index, Qt::DisplayRole)) return absl::StrFormat("%.3f", MakeNumericEntry(index));
+    return MakeStringEntry(index);
+  }
+
+  [[nodiscard]] std::string MakeStringEntry(const Index& index) const {
+    const auto [sfid, column] = index;
+    switch (column) {
+      case Column::kFunctionName:
+        return report_.GetSfidToNames().at(sfid);
       case Column::kIsSignificant:
         return GetPvalue(sfid) < significance_level_ ? "Yes" : "No";
       default:
         ORBIT_UNREACHABLE();
     }
   }
+
```

Comment:
I see that you're trying to avoid code duplication here, but my feeling is that having less intertwined code paths would be more readable - even result in fewer lines of code. So I'm politely asking if you have a strong oppisition against two switch statements for the two roles, even if that meant that a couple of absl::StrFormat calls would be very similar. :see_no_evil:

Reply:
I have to politely oppose. Two similar switches are simple and readable unless you need to spot the difference between the two. For instance, when there's a need to change either. Or when questioning what happens with kIsSignificant column. 

OTOH, I do admit, if there is a need to change only one of the two roles we still have to think what happens with both roles. But I rather see this a good motivation to question once again the fact that we need _only_ one role?

Reply:
Different, but tightly connected. Whatever we choose to sort by, should make sense to the user. If we change sorting role, probably the display role also has to be changed -- otherwise the table might not look sorted to the user.

Reply:
Mainly, because of the significance column. The values we display are yes/no. But they should actually be sorted in the same way we sort pvalues.

Reply:
Oh I see.

```
But they should actually be sorted in the same way we sort pvalues.
```

Coming back to separation of concerns: IMHO this is one thing that should be handled in the proxy model. But this requires subclassing the proxy model. But I don't think I can expext to do that since this would be quite some work for gaining a little more separation.

```
 And we've got a bunch of numeric columns. Wouldn't they be sorted alphabetically then?
```

Ah yeah. True. Yeah, they would be sorted lexicographically if you don't implemented a custom sorting mechanism.